### PR TITLE
Fix headless authentication matching logic for watcher

### PR DESCRIPTION
**Pull request by Joerger**
Update watches on the Headless Authentication resource to properly match events on its corresponding filter.

Closes https://github.com/gravitational/teleport-private/issues/766

### CODE REVIEW

File: api/types/events.go
Comment (sshahcodes) [inline api/types/events.go:151]: Good addition but I since I could not see tests involving `Matches` (pardon my ignorance if it exists) can this break another logic?

File: lib/auth/auth_with_roles.go
Comment (sshahcodes) [inline lib/auth/auth_with_roles.go:1480]: Since the scope of the watcher here is used for authentication, I am curious to understand what scenario may trigger (here and above) the need for `AllowPartialSuccess` ?

Comment (Joerger) [reply lib/auth/auth_with_roles.go:1480]: I can't speak to the exact use case of `AllowPartialSuccess`, but generally it allows a watcher with multiple watch kinds specified to succeed, even if it can only watch a subset of the watch kinds specified. Unauthorized watch kinds do not get watched in this case as they are not added to the list of `validKinds` below, so there is no risk of authz bypass.

https://github.com/gravitational/teleport/blob/a3d89ff6ce231139a3a58f7becb09a122a12b67d/lib/auth/auth_with_roles.go#L1494

File: api/types/events.go
Comment (Joerger) [reply api/types/events.go]: Unfortunately, I think you're right that it could, as there aren't enough tests around the watcher system. I am going to remove this change and leave it out of scope here.

File: lib/auth/auth_with_roles.go
Comment (sshahcodes) [reply lib/auth/auth_with_roles.go:1480]: Noted about the `validKinds`. Even then, I would be inclined to remove `AllowPartialSuccess` altogether if we do not need such partial matches for headless. Your call though!

**Review by flyinghermit — APPROVED** [review]
LGTM as long as we are sure about `AllowPartialSuccess`.

File: lib/auth/auth_with_roles_test.go
Comment (codingllama) [inline lib/auth/auth_with_roles_test.go:5710]:
```suggestion
func TestWatchHeadlessAuthentications_usersCanOnlyWatchThemselves(t *testing.T) {
```

Right?

File: lib/auth/auth_with_roles_test.go
Comment (codingllama) [inline lib/auth/auth_with_roles_test.go:5845]:
```suggestion
					t.Fatalf("Watcher unexpectedly closed with error: %v", watcher.Error())
```

**Review by codingllama — APPROVED** [review]
LGTM, mostly minor comments.

Please update the title and PR description so they better explain the change. The linked issue is private, but also we should provide some context in the PR, apart from the issue itself.

File: lib/auth/auth_with_roles_test.go
Comment (Joerger) [reply lib/auth/auth_with_roles_test.go:5837]: Did a quick search in our code base and it looks like lowerCamel is favored by ~10/1, but TBH I also thought UpperCamelCase was the convention, weird.

Comment (codingllama) [reply lib/auth/auth_with_roles_test.go:5970]: I'm pretty sure UpperCamel was a thing, until it wasn't anymore. Thanks for humoring me.

**Review by codingllama — APPROVED** [review]

**Pull request merged**
